# Project interactive visualization
Using a dataset that your group is consider using for the term project, let's build some meaningful user-driven data visualization. Depending on your dataset this could include:
- Usage of geospatial information
- Building interactive views with widgets
- Organize multiple components into a simple dashboard

Keep these design principles in mind:
While you navigate through this notebook some things to take into consideration:
* Do not add interaction just to add it. Make sure it helps answer a question.
* Use meaningful titles and labels
* Document your code so it's readable and clean. If something does not work, document the issue and explain your best attempt.

In [1]:
# These are most likely the libraries you will use
# Add or remove imports as needed

!pip install hvplot jupyter_bokeh

import pandas as pd
import numpy as np

# Visualization libraries
import plotly.express as px
import plotly.graph_objects as go

# Geospatial / geocoding
from geopy.geocoders import Nominatim
from geopy.exc import GeocoderTimedOut

# Panel
import panel as pn
import panel.widgets as pnw
pn.extension("plotly", "tabulator",)
from panel.interact import interact
import hvplot.pandas

In [5]:
### Load your dataset here

data = pd.read_csv('diabetes_dataset.csv')
data.head()

,age,gender,ethnicity,education_level,income_level,employment_status,smoking_status,alcohol_consumption_per_week,physical_activity_minutes_per_week,diet_score,...,hdl_cholesterol,ldl_cholesterol,triglycerides,glucose_fasting,glucose_postprandial,insulin_level,hba1c,diabetes_risk_score,diabetes_stage,diagnosed_diabetes
0,58,Male,Asian,Highschool,Lower-Middle,Employed,Never,0,215,5.7,...,41,160,145,136,236,6.36,8.18,29.6,Type 2,1
1,48,Female,White,Highschool,Middle,Employed,Former,1,143,6.7,...,55,50,30,93,150,2.00,5.63,23.0,No Diabetes,0
2,60,Male,Hispanic,Highschool,Middle,Unemployed,Never,1,57,6.4,...,66,99,36,118,195,5.07,7.51,44.7,Type 2,1
3,74,Female,Black,Highschool,Low,Retired,Never,0,49,3.4,...,50,79,140,139,253,5.28,9.03,38.2,Type 2,1
4,46,Male,White,Graduate,Middle,Retired,Never,1,109,7.2,...,52,125,160,137,184,12.74,7.20,23.5,Type 2,1


In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 31 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   age                                 100000 non-null  int64  
 1   gender                              100000 non-null  object 
 2   ethnicity                           100000 non-null  object 
 3   education_level                     100000 non-null  object 
 4   income_level                        100000 non-null  object 
 5   employment_status                   100000 non-null  object 
 6   smoking_status                      100000 non-null  object 
 7   alcohol_consumption_per_week        100000 non-null  int64  
 8   physical_activity_minutes_per_week  100000 non-null  int64  
 9   diet_score                          100000 non-null  float64
 10  sleep_hours_per_day                 100000 non-null  float64
 11  screen_time_hours_per_day  

In [8]:
#from google.colab import drive
#drive.mount('/content/drive')

In [9]:
# Write your code here
#data = pd.read_csv('/content/drive/MyDrive/CS133_hw/diabetes_dataset.csv')
#data.head()

## Question 1: Analytical Question
Write a question about your data that can be explored with an interactive plot. A good example would be a question where you can compare different categories. Explain how the plot help answer your question and why you choose that plot type.

Examples:
- Which regions have the highest average values?
- How does one variable compare across time?

QUESTION: How are BMI and fasting glucose related, and how do diabetes stages differ in that space?

**How will the plot help answer my question:**

The plot, specifically the scatter plot allows us to see how BMI vs fasting glucose correlate with one another for each person. Additionally, the diabetes stage being shown as different colors will show whether or not each stages are seperated into different regions, such as having higher fasting glucose and BMI for some stages, or if it overlap a lot. Not only that but the bubble size for weekly physical activity and the information displayed when hovering over a data point help us analysis each dot individually in more details to see whether or not that data point is usual or unusually within its region.

We chose to use a scatter plot for answering this question because since we have two numerical variables for x and y, as well as a categorical variable for grouping, which is the different type of diabetes stage. The scatter plot allows us to see how each person's BMI relates to their fasting glucose level. Furthermore, the scatter plot allows for better visualization of  any trends, clusters, or outliers between the data points.

## Question 2. Create a simple interaction plot
Create a plot, it can be related to your question #1 or a new question, that users can interact with in some meaningful way. Explain in a markdown, how does the interaction add to the analysis.

Example of possible interactions:
*   Hover over information
*   Toogle between groups

In [10]:
#It's hard to visualize the dataset with all 100k row so I kept it to only 20k sample, but I would reccomend changing the random state number
#each time you run if you want to see more than one sample of 20k rows
plot_sample = data.sample(n=20000, random_state=40)

fig_q2 = px.scatter(
    plot_sample,
    x="bmi",
    y="glucose_fasting",
    color="diabetes_stage",
    size="physical_activity_minutes_per_week",
    size_max=14,
    hover_name="gender",
    hover_data={
        "age": True,
        "diagnosed_diabetes": True,
        "hba1c": ":.2f",
        "bmi": ":.1f",
        "glucose_fasting": True,
        "physical_activity_minutes_per_week": True,
    },
    title="BMI vs fasting glucose by diabetes stage (bubble size = weekly activity minutes)",
    labels={
        "bmi": "BMI",
        "glucose_fasting": "Fasting glucose (mg/dL)",
        "diabetes_stage": "Diabetes stage",
        "physical_activity_minutes_per_week": "Activity (min/week)",
    },
    opacity=0.55,
)

fig_q2.update_traces(marker=dict(line=dict(width=0)))
fig_q2.update_layout(legend_title_text="Stage", hovermode="closest")

fig_q2.show()

**How this interaction adds to the analysis:**

The interaction allows us to be able to hover over each data point to shows the age, HbA1c, diagnosis, and activity level for each point so that we can identify and analyze outliers and clusters without overcrowding the chart. In addition, the legend on the right allows us to clicks on different types of diabetes stages to turn the stages on and off to see any overlapping or serparation between the different groups. The zoom interaction allows us to zoom into a smaller region for deeper inspection, especially in regions that are very dense with data points. The pan function allows us to move freely around the zoomed in area, which makes it easier to see a wider range of data points. In conclusion, with such a big dataset of 100k rows, even when we have limited the plot to only randomly include 20k of the 100k datas, the plot still shows very dense regions of datas. Therefore, without the ability to interact with the plot, it would be very difficult for us to explore the pattern between BMI vs fasting glucose by diatbetes stages.

When interacting with the scatter plot, I found that a majority of people with Type 2 diabetes has a fasting glucose level of around 80 to 160 mg/dL with most of the datapoints between 100-130 mg/dL. However, between the BMI level of ~23 and 30, there seems to be more data points clusters for people with Type 2 diabetes. For those who are pre diabetic, the fasting glucose level appears to be mostly between 90 to 120 mg/dL. Similar to those with Typ 2 diabetes, pre diabetic people appear to be mostly around 23-30 for BMI level. For those without diabetes, the cut off for fasting glucose level seems to be right before 100 mg/dL. As for HbA1c, which is the measure for average blood sugar levels, those without diabetes have between mid 4 to mid 5, those who are pre diabetic have between high 5 to mid 6, while those with type 2 diabetes have between 6 to 8 for HbA1c.

## Question 3. Choropleth Planning
Design a choropleth idea using your dataset.
In a markdown cell:
*  Identify the geographic unit you would map (state, county, country, ZIP code, etc.)
*  Identify the variable you would color by
*  Explain if any aggregation or preprocessing is needed
*  Briefly describe what GeoJSON file would be required

You do not need to have a perfect dataset for this question. However, your plan should be realistic.

If your data does not fully support a choropleth, build a prototype table that explains that structure you would need before mapping.

### Proposed Choropleth Design

The dataset lacks direct geographic identifiers. To create a choropleth, I propose mapping the average `diabetes_risk_score` across US states. This would require aggregating individual-level data by hypothetical states to calculate the average risk score per state. A GeoJSON file containing US state boundaries would be necessary for this visualization.

**1. Geographic Unit:** US States

**2. Variable to Color By:** `Average_Diabetes_Risk_Score`

**3. Aggregation/Preprocessing Needed:**
    - **Aggregation:** The `diabetes_risk_score` column would need to be aggregated to calculate the *average* `diabetes_risk_score` for each hypothetical US State. This assumes we could assign each record to a state.
    - **Preprocessing:** Each record in the dataset would hypothetically need to be associated with a US State. Since this information is not present, it would be a critical preprocessing step that is currently missing from the dataset. If this data were available, we would group by state and then calculate the mean `diabetes_risk_score`.

**4. GeoJSON File:** A GeoJSON file containing the boundaries for US states would be required to render the map.

Below is a prototype table demonstrating the required data structure:

| State        | Average_Diabetes_Risk_Score |
| :----------- | :-------------------------- |
| California   | 28.5                        |
| Texas        | 31.2                        |
| New York     | 27.9                        |
| Florida      | 30.1                        |
| Pennsylvania | 29.3                        |

## Question 4. Geospatial Possibility Check

Determine whether your dataset can support a map-based visualization.

In a markdown cell, answer one of the following:
- If **yes**, explain what geographic fields you have and what type of map is appropriate.
- If **no**, explain what is missing and what you would need to create a map.

Write code that inspects or prepares the geographic column(s) you may use.

In [11]:
for n,col in enumerate(data.columns):
    print(n,col)

0 age
1 gender
2 ethnicity
3 education_level
4 income_level
5 employment_status
6 smoking_status
7 alcohol_consumption_per_week
8 physical_activity_minutes_per_week
9 diet_score
10 sleep_hours_per_day
11 screen_time_hours_per_day
12 family_history_diabetes
13 hypertension_history
14 cardiovascular_history
15 bmi
16 waist_to_hip_ratio
17 systolic_bp
18 diastolic_bp
19 heart_rate
20 cholesterol_total
21 hdl_cholesterol
22 ldl_cholesterol
23 triglycerides
24 glucose_fasting
25 glucose_postprandial
26 insulin_level
27 hba1c
28 diabetes_risk_score
29 diabetes_stage
30 diagnosed_diabetes


The initial analysis of the dataset confirmed the lack of direct geographic identifiers such as state, county, or ZIP code. The available columns primarily pertain to individual demographic, lifestyle, and health metrics.

*   **Choropleth Design Proposal:** To visualize diabetes risk geographically, a choropleth mapping the average `diabetes_risk_score` across US states was proposed.
    *   **Geographic Unit:** US States (hypothetically).
    *   **Coloring Variable:** The average `diabetes_risk_score`.
    *   **Aggregation Requirement:** The `diabetes_risk_score` would need to be aggregated to calculate the average for each hypothetical US State.
    *   **Preprocessing Requirement:** A crucial preprocessing step, currently missing, would involve associating each individual record in the dataset with a specific US State.
*   **GeoJSON File Requirement:** A GeoJSON file containing the geographical boundaries for US states is essential for rendering the map. This file should include polygons for each state and properties (e.g., state names or FIPS codes) to link with the aggregated data.
*   **Prototype Data Structure:** A prototype table illustrating the required data structure was provided, featuring a 'State' column and an 'Average\_Diabetes\_Risk\_Score' column, populated with hypothetical values (e.g., California: 28.5, Texas: 31.2).

## Question 5. Geopy / Location Preparation

If your dataset has location names, addresses, cities, states, or countries, use geopy on a sample of your data to geocode locations or validate location information.

If your dataset does not have data that contains location, pick 5 destination you want to visit and use geopy to geocode locations.

Our data does not have locations, so I have picked Tokyo, Paris, Stockholm, Beijing, Sydney

In [12]:
# Your code . . .
# Converted the city destinations into table with city, longitude, and latitude

destinations = {'City': ['Tokyo', 'Paris', 'Stockholm', 'Beijing', 'Sydney']}
geolocator = Nominatim(user_agent="CS133")
longitude = []
latitude = []
df = pd.DataFrame(destinations)
# df
x = {}
for location in destinations['City']:
    x[location] = geolocator.geocode(location)
    longitude.append(x[location].longitude)
    latitude.append(x[location].latitude)


df['longitude'] = longitude
df['latitude'] = latitude
df

,City,longitude,latitude
0,Tokyo,139.763895,35.676860
1,Paris,2.348391,48.853495
2,Stockholm,18.071093,59.325117
3,Beijing,116.391297,39.905714
4,Sydney,151.208285,-33.869844


## Question 6. Panel Widget

Create a Panel Widget that controls something in your analysis such as the ability to choose a column, category, year, etc.

The widget should affect an output such as a plot, table, or summary statistic.

In [13]:
# Your code . . .
# I made a selection and slider widget to see the mean diabetes risk score for each income group by education and weekly alcohol amount
data2 = data
alcohol_widget = pn.widgets.IntSlider(name ='Weekly Alcohol', start = data2['alcohol_consumption_per_week'].min(), end =data2['alcohol_consumption_per_week'].max())
unique_education = list(data['education_level'].unique())
edu_widget = pn.widgets.Select(name = 'Education', options = unique_education)
data2 = data2.interactive()

diabetes_pipeline = (
    data2[(data2['education_level']== edu_widget) & (data2['alcohol_consumption_per_week'] <= alcohol_widget)].groupby(['income_level'])['diabetes_risk_score'].mean().to_frame().reset_index()

)
diabetes_pipeline.hvplot.table()

## Question 7. Mini Dashboard

Build a small Panel dashboard with at least two components. Arrange the components so that it is in a readable layout. Your dashboard should include:
* At least one plot,
* An additional element of your choice such as a widget, table, second plot, etc.

In [16]:

import panel as pn
import hvplot.pandas

idf_q7 = data.interactive()


numeric_options = list(data.select_dtypes(include=[np.number]).columns)
category_options = ['diabetes_stage', 'gender', 'ethnicity', 'income_level']

x_axis = pn.widgets.Select(name='X-Axis (Numeric)', options=numeric_options, value='bmi')
y_axis = pn.widgets.Select(name='Y-Axis (Numeric)', options=numeric_options, value='glucose_fasting')
color_by = pn.widgets.Select(name='Group by Category', options=category_options, value='diabetes_stage')


q7_plot = idf_q7.sample(n=20000, random_state=40).hvplot.scatter(
    x=x_axis,
    y=y_axis,
    by=color_by,
    size=15,
    alpha=0.4,
    title="Comprehensive Healthcare Factor Analysis",
    height=500,
    responsive=True
)


final_dashboard = pn.Column(
    "# Question 7: Interactive Health Explorer",
    "Allow users to compare any two clinical markers across demographics.",
    pn.Row(
        pn.WidgetBox(x_axis, y_axis, color_by, width=250),
        q7_plot.panel()
    )
)

final_dashboard.servable()


Column
    [0] Markdown(str)
    [1] Markdown(str)
    [2] Row
        [0] WidgetBox(width=250)
            [0] Select(name='X-Axis (Numeric)', options=['age', 'alcohol_consumpti...], value='bmi')
            [1] Select(name='Y-Axis (Numeric)', options=['age', 'alcohol_consumpti...], value='glucose_fasting')
            [2] Select(name='Group by Category', options=['diabetes_stage', ...], value='diabetes_stage')
        [1] ParamFunction(function, _pane=HoloViews, defer_load=False)

## Question 8. Reflection

Write a short reflection addressing all of the following:
- Which interactive element was most useful in your notebook?
- What was the hardest part of working with your dataset?
- Did implementing interactive tools help your analysis? Why or why not?
- If you had more time, what would you improve or add?

Which interactive element was most useful in our notebook?

For us, the interactive tools made analyzing the scatter plots much more in-depth. This was especially helpful for sections with many cluttered points, where it’s normally hard to see trends. Being able to hover for details and use the dashboard in Q7 to swap between different variables helped us see the data much more clearly than a static image.

What was the hardest part of working with our dataset?

The hardest part was probably the size of the data. With 100,000 rows, the plots got very crowded and messy. We had to implement sampling strategies to maintain performance.

Did implementing interactive tools help our analysis? Why or why not?

Yes, it helped a lot. Because our dataset has so many different columns (31 in total), we couldn't just make one chart to show everything. Interaction allowed us to swap variables and look at different groups of people quickly, which helped us find patterns.

If we had more time, what would we improve or add?

If we had more time, we would probably add some sort of color or theme to the Question 6 interactive table to make it more appealing to look at. We would also incorporate more columns into that table, move "Education" to its own column to make comparisons easier, and try to link the Q6 table directly to our Q7 plots.